{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/etemadism/MutCrisper/blob/master/MutCrisper_Colab.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "##Install required packages"
      ],
      "metadata": {
        "id": "PBhJLAzLALAd"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%capture\n",
        "#@title Install required packages\n",
        "#packages\n",
        "!pip install Bio\n",
        "!pip install primer3-py\n",
        "!pip install regex\n",
        "!pip install sglearn\n",
        "\n",
        "#!pip install rs3\n",
        "!git clone https://github.com/etemadism/MutCrisper.git\n",
        "!git clone https://github.com/gpp-rnd/rs3.git\n",
        "########libraries ######################\n",
        "import re\n",
        "from Bio.Seq import Seq\n",
        "from tabulate import tabulate\n",
        "import primer3\n",
        "from Bio.Restriction import AllEnzymes\n",
        "import subprocess\n",
        "import os\n",
        "from rs3.rs3.seq import predict_seq"
      ],
      "metadata": {
        "cellView": "form",
        "id": "yGs5TtzRLGO9"
      },
      "execution_count": 25,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Functions\n",
        "###### #######################\n",
        "def validate_sequence(sequence):\n",
        "    \"\"\"\n",
        "    Validates the input DNA sequence.\n",
        "    \"\"\"\n",
        "    valid_chars = set(\"ACTG\")\n",
        "    if not set(sequence.upper()).issubset(valid_chars):\n",
        "        raise ValueError(\"Invalid DNA sequence. Only A, C, T, G characters are allowed.\")\n",
        "\n",
        "def validate_nucleotide_position(position, sequence):\n",
        "    \"\"\"\n",
        "    Validates the nucleotide position.\n",
        "    \"\"\"\n",
        "    if position < 1 or position > len(sequence):\n",
        "        raise ValueError(\"Invalid nucleotide position. Please enter a valid position within the sequence.\")\n",
        "\n",
        "def get_search_strand(sequence, position):\n",
        "    \"\"\"\n",
        "    Determines the search strand based on the nucleotide at the position of interest.\n",
        "    \"\"\"\n",
        "    nucleotide = sequence[position - 1]\n",
        "    return \"sense\" if nucleotide in [\"A\", \"C\"] else \"antisense\"\n",
        "\n",
        "def calculate_cg_percent(sequence):\n",
        "    \"\"\"\n",
        "    Calculates the CG percent of a given sequence.\n",
        "    \"\"\"\n",
        "    cg_count = sequence.count(\"C\") + sequence.count(\"G\")\n",
        "    return (cg_count / len(sequence)) * 100\n",
        "\n",
        "def find_pam_positions(sequence, search_strand):\n",
        "    \"\"\"\n",
        "    Finds all PAM positions on the selected strand.\n",
        "    \"\"\"\n",
        "    if search_strand == \"sense\":\n",
        "        pam_motif = re.compile(\"(?=AGG|TGG|CGG|GGG)\")\n",
        "    else:\n",
        "        pam_motif = re.compile(\"(?=AGG|TGG|CGG|GGG)\") #?=CCA|CCT|CCC|CCG #\n",
        "    return [match.start() for match in pam_motif.finditer(sequence)]\n",
        "\n",
        "\n",
        "\n",
        "def find_sorted_efficacies(BE_profile, nucleotide_pos):\n",
        "    efficacies = []\n",
        "\n",
        "    for cell, enzyme, *efficacy_values in BE_profile:\n",
        "        efficacy = efficacy_values[pos_grna]\n",
        "        nucleotide = sequence[nucleotide_pos - 1]\n",
        "        if nucleotide == 'A':\n",
        "            if enzyme in ['ABE', 'ABE-CP']:\n",
        "                if efficacy > 0:\n",
        "                    efficacies.append((cell, enzyme, efficacy))\n",
        "        else:\n",
        "            if enzyme in ['BE4', 'BE4-CP', 'evoA-BE4', 'AID-BE4', 'eA3A-BE4', 'CDA-BE4']:\n",
        "                if efficacy > 0:\n",
        "                    efficacies.append((cell, enzyme, efficacy))\n",
        "\n",
        "    sorted_efficacies = sorted(efficacies, reverse=True)\n",
        "    return sorted_efficacies\n",
        "\n",
        "\n",
        "def get_re_wild(wild, mut):\n",
        "    enzymes = AllEnzymes\n",
        "    for enzyme in enzymes:\n",
        "        if enzyme.site in wild and enzyme.site not in mut:\n",
        "            print(enzyme, enzyme.site)\n",
        "    print(\" \")\n",
        "\n",
        "def get_re_mut(wild, mut):\n",
        "    enzymes = AllEnzymes\n",
        "    for enzyme in enzymes:\n",
        "        if enzyme.site in mut and enzyme.site not in wild:\n",
        "            print(enzyme, enzyme.site)\n",
        "    print(\" \")\n",
        "\n",
        "def generate_mutant_sequence(sequence, nucleotide_pos):\n",
        "    pos9 = nucleotide_pos - 9\n",
        "    pos8 = nucleotide_pos + 8\n",
        "\n",
        "    wild = sequence[pos9:pos8]\n",
        "    #mut = wild[:8] + str(Seq(wild[8]).complement()) + wild[9:]\n",
        "    # if search_strand== \"sense\":\n",
        "    #     sequence=sequence\n",
        "    # else:\n",
        "    #     sequence= revcomp_sequence\n",
        "\n",
        "    #wild = sequence[pos9:pos8]\n",
        "    nuc_mut = Seq(wild[8])\n",
        "\n",
        "    if nuc_mut == \"A\":\n",
        "        mut = wild[:8] + \"G\" + wild[9:]\n",
        "    elif nuc_mut == \"C\":\n",
        "        mut = wild[:8] + \"T\" + wild[9:]\n",
        "    else:\n",
        "        mut = wild  # No mutation needed for other nucleotides\n",
        "\n",
        "    print(\"This is the wild type sequence:   \", wild)\n",
        "    print(\"This is the mutant type sequence: \", mut)\n",
        "\n",
        "    print(\"These enzymes will potentially cut the wild sequence, but not mutant sequence.\")\n",
        "    get_re_wild(wild, mut)\n",
        "\n",
        "\n",
        "    print(\"These enzymes will potentially cut the mutant sequence, but not wild sequence.\")\n",
        "    get_re_mut(wild, mut)\n",
        "\n",
        "# Define the path to the CFD_Scoring directory\n",
        "cfd_scoring_path = '/content/MutCrisper/CFD_Scoring'\n",
        "\n",
        "# Function for CFD score calculation\n",
        "def cfd_score_calculator(wt, off):\n",
        "    # Store the current working directory\n",
        "    current_dir = os.getcwd()\n",
        "\n",
        "    try:\n",
        "        # Change the current working directory temporarily for the subprocess\n",
        "        os.chdir(cfd_scoring_path)\n",
        "\n",
        "        command = f\"python cfd_score_calculator.py --wt {wt} --off {off}\"\n",
        "        process = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
        "        stdout, stderr = process.communicate()\n",
        "\n",
        "        # Check if the subprocess was executed successfully\n",
        "        if process.returncode != 0:\n",
        "            raise RuntimeError(f\"Error executing cfd_score_calculator.py: {stderr.decode()}\")\n",
        "\n",
        "        # Extract the numerical value from the output\n",
        "        output = stdout.decode().strip()\n",
        "        match = re.search(r\"(\\d+\\.\\d+)\", output)\n",
        "        if match:\n",
        "            score = float(match.group(1))\n",
        "        else:\n",
        "            raise ValueError(\"Could not extract CFD score from the output.\")\n",
        "\n",
        "        return round(score, 3)\n",
        "    finally:\n",
        "        # Restore the original working directory\n",
        "        os.chdir(current_dir)"
      ],
      "metadata": {
        "cellView": "form",
        "id": "x27QnQorBJvc"
      },
      "execution_count": 26,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Base editor activity profile\n",
        "\n",
        "####\n",
        "######################################################################################################\n",
        "##################Base editor activity profiles#######################################################\n",
        "##################Reference: Mandana Arbab et al. Cell. 2020. doi:10.1016/j.cell.2020.05.037##########\n",
        "######################################################################################################\n",
        "BE_profile = [[\n",
        "  'mESC', 'ABE', 0.0, 0.0, 0.0, 37.4, 85.1, 100.0, 77.7, 43.1, 0.0, 0.0, 0.0,\n",
        "  0.0, 0.0, 0.0, 0.0, 0, 0, 0, 0, 0], [\n",
        "  'mESC', 'ABE-CP', 0.0, 0.0, 41.0, 83.9, 91.9, 100.0, 78.8, 63.1, 52.6, 50.5,\n",
        "  0.0, 0.0, 0.0, 0.0, 0.0, 0, 0, 0, 0, 0], [\n",
        "  'mESC', 'BE4', 0.0, 0.0, 31.8, 63.2, 90.3, 100.0, 87.0, 62.0, 31.4, 0.0, 0.0,\n",
        "  0.0, 0.0, 0.0, 0.0, 0, 0, 0, 0, 0], [\n",
        "  'mESC', 'BE4-CP', 0.0, 42.5, 47.9, 69.3, 80.8, 92.1, 97.0, 100.0, 97.3, 91.7,\n",
        "  76.8, 65.9, 59.6, 38.5, 34.7, 0, 0, 0, 0, 0], [\n",
        "  'mESC', 'evoA-BE4', 0.0, 44.4, 72.1, 97.3, 100.0, 98.9, 86.7, 63.6, 36.5, 0.0,\n",
        "  0.0, 0.0, 0.0, 0.0, 0.0, 0, 0, 0, 0, 0], [\n",
        "  'mESC', 'AID-BE4', 46.1, 62.4, 82.4, 87.2, 90.2, 95.9, 100.0, 86.6, 76.9, 56.5,\n",
        "  32.3, 0.0, 0.0, 0.0, 0.0, 0, 0, 0, 0, 0], [\n",
        "  'mESC', 'eA3A-BE4', 0.0, 0.0, 0.0, 41.6, 83.6, 100.0, 98.5, 75.7, 35.9, 0.0, 0.0,\n",
        "  0.0, 0.0, 0.0, 0.0, 0, 0, 0, 0, 0], [\n",
        "  'mESC', 'CDA-BE4', 88.4, 98.9, 97.8, 99.4, 100.0, 93.5, 90.4, 68.8, 40.1, 0.0,\n",
        "  0.0, 0.0, 0.0, 0.0, 0.0, 0, 0, 0, 0, 0], [\n",
        "  'HEK293T', 'ABE', 0.0, 0.0, 0.0, 54.2, 74.2, 100.0, 86.1, 50.2, 0.0, 0.0, 0.0, 0.0,\n",
        "  0.0, 0.0, 0.0, 0, 0, 0, 0, 0], [\n",
        "  'HEK293T', 'ABE-CP', 0.0, 0.0, 39.4, 59.4, 75.7, 77.4, 100.0, 76.1, 58.3, 52.4, 44.6,\n",
        "  0.0, 0.0, 0.0, 0.0, 0, 0, 0, 0, 0], [\n",
        "  'HEK293T', 'BE4', 0.0, 0.0, 36.1, 68.4, 100.0, 90.7, 72.2, 59.0, 37.3, 0.0, 0.0, 0.0,\n",
        "  0.0, 0.0, 0.0, 0, 0, 0, 0, 0], [\n",
        "  'HEK293T', 'BE4-CP', 0.0, 35.8, 50.4, 75.9, 81.3, 86.8, 83.7, 84.6, 100.0, 92.9, 92.0,\n",
        "  59.5, 56.1, 56.2, 37.9, 0, 0, 0, 0, 0], [\n",
        "  'HEK293T', 'evoA-BE4', 0.0, 39.7, 70.8, 97.7, 100.0, 94.8, 79.6, 67.1, 39.0, 0.0, 0.0,\n",
        "  0.0, 0.0, 0.0, 0.0, 0, 0, 0, 0, 0], [\n",
        "  'HEK293T', 'AID-BE4', 49.0, 62.0, 93.1, 87.9, 82.9, 100.0, 82.9, 97.0, 75.1, 65.6, 33.8,\n",
        "  33.6, 0.0, 0.0, 0.0, 0, 0, 0, 0, 0], [\n",
        "  'HEK293T', 'eA3A-BE4', 0.0, 0.0, 0.0, 35.8, 85.2, 100.0, 76.4, 73.7, 35.7, 0.0, 0.0, 0.0,\n",
        "  0.0, 0.0, 0.0, 0, 0, 0, 0, 0], [\n",
        "  'HEK293T', 'CDA-BE4', 80.8, 79.3, 84.6, 100.0, 83.3, 89.0, 80.0, 73.6, 57.3, 37.7, 0.0,\n",
        "  0.0, 0.0, 0.0, 0.0, 0, 0, 0, 0, 0], [\n",
        "  'HEK293T', 'eA3A-BE5', 0.0, 0.0, 0.0, 42.1, 67.9, 95.9, 100.0, 65.6, 32.1, 0.0, 0.0, 0.0,\n",
        "  0.0, 0.0, 0.0, 0, 0, 0, 0, 0], [\n",
        "  'U2OS', 'ABE', 0.0, 0.0, 0.0, 38.9, 100.0, 97.1, 83.5, 43.4, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,\n",
        "  0.0, 0, 0, 0, 0, 0], [\n",
        "  'U2OS', 'ABE-CP', 0.0, 0.0, 34.1, 71.5, 83.0, 100.0, 90.1, 82.2, 55.1, 49.1, 44.4, 0.0, 0.0,\n",
        "  0.0, 0.0, 0, 0, 0, 0, 0], [\n",
        "  'U2OS', 'BE4', 0.0, 0.0, 41.1, 66.3, 100.0, 99.0, 93.2, 56.1, 40.3, 0.0, 0.0, 0.0, 0.0, 0.0,\n",
        "  0.0, 0, 0, 0, 0, 0], [\n",
        "  'U2OS', 'BE4-CP', 0.0, 34.5, 57.0, 64.9, 90.5, 82.7, 82.9, 100.0, 94.1, 94.9, 79.9, 81.0,\n",
        "  50.0, 32.0, 0.0, 0, 0, 0, 0, 0], [\n",
        "  'U2OS', 'evoA-BE4', 0.0, 32.1, 64.3, 92.4, 91.7, 100.0, 79.9, 61.2, 45.7, 0.0, 0.0, 0.0, 0.0,\n",
        "  0.0, 0.0, 0, 0, 0, 0, 0], [\n",
        "  'U2OS', 'AID-BE4', 49.0, 62.0, 93.1, 87.9, 82.9, 100.0, 82.9, 97.0, 75.1, 65.6, 33.8, 33.6,\n",
        "  0.0, 0.0, 0.0, 0, 0, 0, 0, 0], [\n",
        "  'U2OS', 'eA3A-BE4', 0.0, 0.0, 0.0, 0.0, 74.7, 100.0, 99.9, 65.7, 38.0, 0.0, 0.0, 0.0, 0.0, 0.0,\n",
        "  0.0, 0, 0, 0, 0, 0], [\n",
        "  'U2OS', 'CDA-BE4', 80.8, 79.3, 84.6, 100.0, 83.3, 89.0, 80.0, 73.6, 57.3, 37.7, 0.0, 0.0, 0.0,\n",
        "  0.0, 0.0, 0, 0, 0, 0, 0]]\n",
        "########################################################################################################\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "nvzAu6oBBn_j"
      },
      "execution_count": 27,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "##Inputs"
      ],
      "metadata": {
        "id": "vhiq7r_dHK1C"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Please enter a DNA sequence and a position of the nucleotide of interest.\n",
        "\n",
        "\n",
        "sequence = 'GGGCATCATCAATTTCGAGCAGAAGGCAAGGGCTGGGACGGAGGC' #@param {type:\"string\"}\n",
        "#sequence= input(\"Please enter a DNA sequence(with at least 20nt flanking in upstream and downstream of position of the nucleotide of interest): \")\n",
        "\n",
        "nucleotide_pos = 20 # @param {type:\"number\"}"
      ],
      "metadata": {
        "id": "7HWaxddnEuAb",
        "cellView": "form"
      },
      "execution_count": 14,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "##Run MutCrisper"
      ],
      "metadata": {
        "id": "_OCJr1-dAhMo"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Run MutCrisper\n",
        "# Validate sequence\n",
        "validate_sequence(sequence)\n",
        "sequence = sequence.upper()\n",
        "org_sequence=sequence\n",
        "\n",
        "# Validate position\n",
        "validate_nucleotide_position(nucleotide_pos, sequence)\n",
        "\n",
        "\n",
        "org_nucleotide_pos=nucleotide_pos\n",
        "\n",
        "###name of nucleotide of interest\n",
        "org_nucleotide = sequence[org_nucleotide_pos - 1]\n",
        "\n",
        "###mutation type for ARMS PCR\n",
        "# Mutation type for ARMS PCR\n",
        "if org_nucleotide == \"A\":\n",
        "    mut_nucleotide = \"G\"\n",
        "elif org_nucleotide == \"C\":\n",
        "    mut_nucleotide = \"T\"\n",
        "elif org_nucleotide == \"T\":\n",
        "    mut_nucleotide = \"C\"\n",
        "elif org_nucleotide == \"G\":\n",
        "    mut_nucleotide = \"A\"\n",
        "else:\n",
        "\n",
        "    mut_nucleotide = None  # This is a wrong nt letter.\n",
        "\n",
        "\n",
        "\n",
        "# Create a pattern to match the nucleotide of interest\n",
        "nucleotide_pattern = re.compile(f\"^(?P<before>.{{0,{nucleotide_pos-1}}})(?P<nucleotide>.)(?P<after>.*)$\")\n",
        "\n",
        "# Create a formatted string with the bold nucleotide position\n",
        "formatted_sequence = nucleotide_pattern.sub(\"\\g<before>\\033[1m\\g<nucleotide>\\033[0m\\g<after>\", sequence)\n",
        "\n",
        "\n",
        "# Print the formatted sequence\n",
        "print(f\"The position of interest is shown as bold: {formatted_sequence}\")\n",
        "\n",
        "# Determine the search strand based on the nucleotide at the position of interest\n",
        "search_strand = get_search_strand(sequence, nucleotide_pos)\n",
        "print(f\"You need sgRNA for {search_strand} sequence.\\n\")\n",
        "\n",
        "\n",
        "\n",
        "# Reverse complement the sequence if it's on the antisense strand\n",
        "if search_strand == \"antisense\":\n",
        "    revcomp_sequence = str(Seq(sequence).reverse_complement())\n",
        "\n",
        "    nucleotide_pos= len(sequence)- (nucleotide_pos - 1)\n",
        "    nucleotide_pattern = re.compile(f\"(?P<before>.{{0,{nucleotide_pos-1}}})(?P<nucleotide>.)(?P<after>.*)\")\n",
        "    formatted_sequence = nucleotide_pattern.sub(\"\\g<before>\\033[1m\\g<nucleotide>\\033[0m\\g<after>\", revcomp_sequence)\n",
        "    print(f\"Here is reverse complement:                {formatted_sequence}\")\n",
        "if search_strand == \"sense\":\n",
        "    sequence=sequence\n",
        "else:\n",
        "    sequence=revcomp_sequence\n",
        "\n",
        "# Find all PAM positions on the selected strand\n",
        "pam_positions = find_pam_positions(sequence if search_strand == \"sense\" else revcomp_sequence, search_strand)\n",
        "\n",
        "\n",
        "\n",
        "# Find all gRNAs\n",
        "sgRNA_list = []\n",
        "\n",
        "for pos in pam_positions:\n",
        "    if (search_strand == \"sense\" and 0 <= (pos - nucleotide_pos) < 20) or (search_strand == \"antisense\" and 0 <= (pos - nucleotide_pos) < 20):\n",
        "        pPAM= pos + 3\n",
        "        pPAM6= pos + 6\n",
        "        gRNA_start = pos - 20\n",
        "        gRNA_start4= pos - 24\n",
        "        gPAM= sequence[gRNA_start:pPAM]\n",
        "        qPAM30=sequence[gRNA_start4:pPAM6]\n",
        "\n",
        "        if search_strand == \"sense\":\n",
        "\n",
        "            gRNA = sequence[gRNA_start:pos]\n",
        "            gRNA30 = sequence[gRNA_start4:pPAM6]\n",
        "\n",
        "        else:\n",
        "\n",
        "            gRNA = revcomp_sequence[gRNA_start:pos]\n",
        "            gRNA30 = revcomp_sequence[gRNA_start4:pPAM6]\n",
        "\n",
        "        # Check if gRNA is empty\n",
        "        if not gRNA:\n",
        "            print(\"Error: gRNA not found.\")\n",
        "            continue\n",
        "\n",
        "        cg_percent = calculate_cg_percent(gRNA)\n",
        "\n",
        "        nucleotide_pos = nucleotide_pos  # Specify the nucleotide position of interest\n",
        "        end = pos + 1  #pos - 1\n",
        "        start = end - 20\n",
        "        pos_grna = 20 - (end - nucleotide_pos)\n",
        "        wt = gPAM  # Assign gRNA to wt\n",
        "\n",
        "        # Create a list of nucleotides in the gRNA with the specified replacements\n",
        "        off_list = [nt if i != pos_grna else 'G' if nt == 'A' else 'T' if nt == 'C' else nt for i, nt in enumerate(gPAM)]\n",
        "\n",
        "        # Join the modified nucleotides to form the off-gRNA sequence\n",
        "        off = ''.join(off_list)\n",
        "        # calculate cfd score\n",
        "        cfd_score= cfd_score_calculator(wt, off)\n",
        "\n",
        "\n",
        "        sorted_efficacies = find_sorted_efficacies(BE_profile, nucleotide_pos)\n",
        "        tm= primer3.calc_tm(gRNA)\n",
        "        Tm = round(tm, 2)\n",
        "        hp=primer3.calc_hairpin(gRNA)\n",
        "        Hairpin=hp.structure_found\n",
        "\n",
        "\n",
        "        ###RS3 score\n",
        "        #context_seqs = [gRNA30]\n",
        "        # RS3_Hsu_score= predict_seq(context_seqs, sequence_tracr='Hsu2013')\n",
        "        # RS3_Chen_score= predict_seq(context_seqs, sequence_tracr='Chen2013')\n",
        "\n",
        "\n",
        "        sgRNA_list.append([gRNA +' NGG', sorted_efficacies, cfd_score, cg_percent, Tm, Hairpin ]) #RS3_Hsu_score,RS3_Chen_score,\n",
        "\n",
        "print(\"                                 Table of sgRNA sequences and suggested conditions \")\n",
        "print(\"----------------------------------------------------------------------------------------------------------------\")\n",
        "\n",
        "headers = [\"sgRNA Sequence\", \"Cell     Enzyme  Efficacy(%)\", \"CFD specificity score\", \"GC(%)\",  'Tm(°C)', \"Hairpin?\"]#\"GC(%)\",\"RS3 Score(Hsu)\", \"RS3 Score(Chen)\",\n",
        "for row in sgRNA_list:\n",
        "    row[1] = tabulate(row[1], tablefmt=\"plain\")  # Format sorted efficacies as a table\n",
        "print(tabulate(sgRNA_list, headers=headers))\n",
        "\n",
        "print(\"----------------------------------------------------------------------------------------------------------------\")\n",
        "\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "pHH9kLo0OZ6e",
        "outputId": "0f52bf53-f318-45cc-ab80-f60189bfb249",
        "cellView": "form"
      },
      "execution_count": 18,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "The position of interest is shown as bold: GGGCATCATCAATTTCGAG\u001b[1mC\u001b[0mAGAAGGCAAGGGCTGGGACGGAGGC\n",
            "You need sgRNA for sense sequence.\n",
            "\n",
            "                                 Table of sgRNA sequences and suggested conditions \n",
            "----------------------------------------------------------------------------------------------------------------\n",
            "sgRNA Sequence            Cell     Enzyme  Efficacy(%)      CFD specificity score    GC(%)    Tm(°C)  Hairpin?\n",
            "------------------------  ------------------------------  -----------------------  -------  --------  ----------\n",
            "CATCATCAATTTCGAGCAGA NGG                                                    0.467       40     51.65  False\n",
            "TCAATTTCGAGCAGAAGGCA NGG  mESC     BE4-CP   65.9                            0.538       45     55.68  True\n",
            "                          U2OS     BE4-CP   81\n",
            "                          U2OS     AID-BE4  33.6\n",
            "                          HEK293T  BE4-CP   59.5\n",
            "                          HEK293T  AID-BE4  33.6\n",
            "CAATTTCGAGCAGAAGGCAA NGG  mESC     BE4-CP   76.8                            0.308       45     54.86  True\n",
            "                          mESC     AID-BE4  32.3\n",
            "                          U2OS     BE4-CP   79.9\n",
            "                          U2OS     AID-BE4  33.8\n",
            "                          HEK293T  BE4-CP   92\n",
            "                          HEK293T  AID-BE4  33.8\n",
            "TTCGAGCAGAAGGCAAGGGC NGG  mESC     evoA-BE4   86.7                          0.812       60     61     True\n",
            "                          mESC     eA3A-BE4   98.5\n",
            "                          mESC     CDA-BE4    90.4\n",
            "                          mESC     BE4-CP     97\n",
            "                          mESC     BE4        87\n",
            "                          mESC     AID-BE4   100\n",
            "                          U2OS     evoA-BE4   79.9\n",
            "                          U2OS     eA3A-BE4   99.9\n",
            "                          U2OS     CDA-BE4    80\n",
            "                          U2OS     BE4-CP     82.9\n",
            "                          U2OS     BE4        93.2\n",
            "                          U2OS     AID-BE4    82.9\n",
            "                          HEK293T  evoA-BE4   79.6\n",
            "                          HEK293T  eA3A-BE4   76.4\n",
            "                          HEK293T  CDA-BE4    80\n",
            "                          HEK293T  BE4-CP     83.7\n",
            "                          HEK293T  BE4        72.2\n",
            "                          HEK293T  AID-BE4    82.9\n",
            "TCGAGCAGAAGGCAAGGGCT NGG  mESC     evoA-BE4   98.9                          0.929       60     61.92  True\n",
            "                          mESC     eA3A-BE4  100\n",
            "                          mESC     CDA-BE4    93.5\n",
            "                          mESC     BE4-CP     92.1\n",
            "                          mESC     BE4       100\n",
            "                          mESC     AID-BE4    95.9\n",
            "                          U2OS     evoA-BE4  100\n",
            "                          U2OS     eA3A-BE4  100\n",
            "                          U2OS     CDA-BE4    89\n",
            "                          U2OS     BE4-CP     82.7\n",
            "                          U2OS     BE4        99\n",
            "                          U2OS     AID-BE4   100\n",
            "                          HEK293T  evoA-BE4   94.8\n",
            "                          HEK293T  eA3A-BE4  100\n",
            "                          HEK293T  CDA-BE4    89\n",
            "                          HEK293T  BE4-CP     86.8\n",
            "                          HEK293T  BE4        90.7\n",
            "                          HEK293T  AID-BE4   100\n",
            "GCAGAAGGCAAGGGCTGGGA NGG  mESC     evoA-BE4  44.4                           0.909       65     62.88  True\n",
            "                          mESC     CDA-BE4   98.9\n",
            "                          mESC     BE4-CP    42.5\n",
            "                          mESC     AID-BE4   62.4\n",
            "                          U2OS     evoA-BE4  32.1\n",
            "                          U2OS     CDA-BE4   79.3\n",
            "                          U2OS     BE4-CP    34.5\n",
            "                          U2OS     AID-BE4   62\n",
            "                          HEK293T  evoA-BE4  39.7\n",
            "                          HEK293T  CDA-BE4   79.3\n",
            "                          HEK293T  BE4-CP    35.8\n",
            "                          HEK293T  AID-BE4   62\n",
            "----------------------------------------------------------------------------------------------------------------\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Set Primer1\n",
        "# Suppress the output of the %cd command\n",
        "%cd -q /content/MutCrisper/primer1/primer1_source\n",
        "!rm *.class\n",
        "\n",
        "##Compile the Java source code\n",
        "!javac doPrimer.java\n",
        "\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "znswO6aYYfbp"
      },
      "execution_count": 20,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Restriction enzymes(if any) for RFLP\n",
        "\n",
        "#####Generate mutant sequence and find possible RFLP restriction enzyms\n",
        "generate_mutant_sequence(sequence, nucleotide_pos)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "cellView": "form",
        "id": "FT_ZVPRWRy8y",
        "outputId": "660f1f09-3cb7-4550-aeed-009fda98d3df"
      },
      "execution_count": 21,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "This is the wild type sequence:    ATTTCGAGCAGAAGGCA\n",
            "This is the mutant type sequence:  ATTTCGAGTAGAAGGCA\n",
            "These enzymes will potentially cut the wild sequence, but not mutant sequence.\n",
            "Esp3007I CAGAAG\n",
            " \n",
            "These enzymes will potentially cut the mutant sequence, but not wild sequence.\n",
            " \n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Please specify the parameters for ARMS PCR primer design(http://primer1.soton.ac.uk/primer1.html):\n",
        "Optimum_primer_size = 28 #@param {type:\"number\"}\n",
        "Maximum_primer_size= 30 #@param {type:\"number\"}\n",
        "Minimum_primer_size= 26 #@param {type:\"number\"}\n",
        "Optimum_inner_product_size = 200 #@param {type:\"number\"}\n",
        "Maximum_inner_product_size= 300 #@param {type:\"number\"}\n",
        "Minimum_inner_product_size= 50 #@param {type:\"number\"}\n",
        "Maximum_relative_size = 1.5 #@param {type:\"number\"}\n",
        "Minimum_relative_size= 1.1 #@param {type:\"number\"}\n",
        "Optimum_primer_Tm= 65 #@param {type:\"number\"}\n",
        "Maximum_primer_Tm= 80 #@param {type:\"number\"}\n",
        "Minimum_primer_Tm= 50 #@param {type:\"number\"}\n",
        "Maximum_primer_GC= 80 #@param {type:\"number\"}\n",
        "Minimum_primer_GC= 10 #@param {type:\"number\"}\n",
        "Maximum_complementarity= 8.0 #@param {type:\"number\"}\n",
        "Maximum_3_complementarity= 3.0 #@param {type:\"number\"}\n",
        "Salt_concentration_mM= 50 #@param {type:\"number\"}\n",
        "Annealing_primer_concentration_mM= 50 #@param {type:\"number\"}\n",
        "Number_of_outputs= 10 #@param {type:\"number\"}\n",
        "\n",
        "!java -classpath /content/MutCrisper/primer1/primer1_source doPrimer '{org_sequence} {org_nucleotide_pos} {org_nucleotide} {mut_nucleotide} {Optimum_primer_size} {Maximum_primer_size} {Minimum_primer_size} {Optimum_inner_product_size} {Maximum_inner_product_size} {Minimum_inner_product_size} {Maximum_relative_size} {Minimum_relative_size} {Optimum_primer_Tm} {Maximum_primer_Tm} {Minimum_primer_Tm} {Maximum_primer_GC} {Minimum_primer_GC} {Maximum_complementarity} {Maximum_3_complementarity} {Salt_concentration_mM} {Annealing_primer_concentration_mM} {Number_of_outputs}'\n",
        "\n",
        "\n",
        "\n",
        "print ('''\n",
        "\n",
        "******************************************************Disclaimer:***********************************************\n",
        "This code is for educational purposes only. Use it at your own risk. The author of this code makes no warranties\n",
        "about the accuracy or suitability of this code for any purpose. By using this code, you agree to hold the author\n",
        "harmless from any damages that may arise from its use.\n",
        "''')\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "cellView": "form",
        "id": "Q7-3Bt2yKsUo",
        "outputId": "70c79426-eb76-4b6d-c8e0-6a656683fd55"
      },
      "execution_count": 30,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Sequence before the SNP too short to design a primer\n",
            "\n",
            "\n",
            "******************************************************Disclaimer:***********************************************\n",
            "This code is for educational purposes only. Use it at your own risk. The author of this code makes no warranties\n",
            "about the accuracy or suitability of this code for any purpose. By using this code, you agree to hold the author\n",
            "harmless from any damages that may arise from its use.\n",
            "\n"
          ]
        }
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "collapsed_sections": [
        "PBhJLAzLALAd"
      ],
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}