<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1">
<meta http-equiv="Expires" content="javascript:returnExpiry(5)">
<title>Help information</title>

<SCRIPT LANGUAGE="Javascript">
//<! -- HIDE FROM OTHER BROWSERS

function returnExpiry(days) {
    var expiryDate = new Date();
    expiryDate.setDate(expiryDate.getDate() + days);
    return(expiryDate.toGMTString());
}

//STOP HIDING FROM OTHER BROWSERS -->
</SCRIPT>

</head>

<Body BGCOLOR=#CCEEFF TEXT=#000000>

<h3><a name="top">Tetra-primer ARMS PCR</a></h3>
<ul>
<li><a href="#source_sequence">Source sequence</a></li>
<li><a href="#SNP_position">Position of SNP from start of sequence</a></li>
<li><a href="#allele1">Allele1</a></li>
<li><a href="#allele2">Allele2</a></li>
<li><a href="#optprimersize">Optimum primer size</a></li>
<li><a href="#maxprimersize">Maximum primer size</a></li>
<li><a href="#minprimersize">Minimum primer size</a></li>
<li><a href="#optproductsize">Optimum (inner) product size</a></li>
<li><a href="#maxproductsize">Maximum (inner) product size</a></li>
<li><a href="#minproductsize">Minimum (inner) product size</a></li>
<li><a href="#maxproductsizedif">Maximum relative size difference of two inner products</a></li>
<li><a href="#minproductsizedif">Minimum relative size difference of two inner products</a></li>
<li><a href="#optprimerTm">Optimum primer Tm</a></li>
<li><a href="#maxprimerTm">Maximum primer Tm</a></li>
<li><a href="#minprimerTm">Minimum primer Tm</a></li>
<li><a href="#maxprimerGC">Maximum primer GC%</a></li>
<li><a href="#minprimerGC">Minimum primer GC%</a></li>
<li><a href="#maxcom">Maximum complementarity</a></li>
<li><a href="#max3com">Maximum 3' complementarity</a></li>
<li><a href="#saltconc">Salt concentration (mM)</a></li>
<li><a href="#primerconc">Annealing primer concentration (nM)</a></li>
<li><a href="#numoutput">Number of outputs</a></li>
</ul>

<h4><a name="source_sequence">Source sequence</a></h4>
The sequence from which to choose primers. The sequence must be presented 5'-3'. The bases may be upper or lower
case. Whitespace characters and digits are ignored, and any other unrecognized character, other than A, T, C, G,
is treated as N.<br>

<h4><a name="SNP_position">Position of SNP from start of sequence</a></h4>
The position of the SNP allele in the source sequence.<br>

<h4><a name="allele1">Allele1</a></h4>
The SNP allele as indicated in the source sequence.<br>

<h4><a name="allele2">Allele2</a></h4>
The other SNP allele (the one not in the source sequence).<br>

<h4><a name="optprimersize">Optimum primer size</a></h4>
The value specifies the optimum length of the primers (in bases). The program will pick primers close
to this length.<br>

<h4><a name="maxprimersize">Maximum primer size</a></h4>
The value specifies the maximum acceptable length of the primers (in bases). This parameter should not
be larger than 35. This limit is governed by maximum oligo size for which the program's melting-
temperature is valid<br>

<h4><a name="minprimersize">Minimum primer size</a></h4>
The value specifies the minimum acceptable length of the primers (in bases).<br>

<h4><a name="optproductsize">Optimum (inner) product size</a></h4>
The value specifies the optimum length of the PCR products, which are between one allele specific primer
and its corresponding outer primer (on the opposite strand). Therefore, this value does not apply to the
PCR products between the two outer primers.<br>

<h4><a name="maxproductsize">Maximum (inner) product size</a></h4>
The value specifies the maximum length of the PCR products, which are between one allele specific primer
and its corresponding outer primer (on the opposite strand). Therefore, this value does not apply to the
PCR products between the two outer primers.<br>

<h4><a name="minproductsize">Minimum (inner) product size</a></h4>
The value specifies the minimum length of the PCR products, which are between one allele specific primer
and its corresponding outer primer (on the opposite strand). Therefore, this value does not apply to the
PCR products between the two outer primers.<br>

<h4><a name="maxproductsizedif">Maximum relative size difference of two inner products</a></h4>
The value specifies the maximum size difference between the two allele specific PCR products. The size
difference value is the result of the division of the longer product by the shorter one. Floating point
values bigger than 1.0 are usually expected.<br>
 
<h4><a name="minproductsizedif">Minimum relative size difference of two inner products</a></h4>
The value specifies the minimum size difference between the two allele specific PCR products. The size
difference value is the result of the division of the longer product by the shorter one. Floating point
values bigger than 1.0 are usually expected so as to seperate the two prodcuts on gel according to their
size difference.<br>
 
<h4><a name="optprimerTm">Optimum primer Tm</a></h4>
Optimum melting temperature(Celsius) for a primer oligo. The program will try to pick primers with
melting temperatures are close to this temperature.<br>

<h4><a name="maxprimerTm">Maximum primer Tm</a></h4>
Maximum acceptable melting temperature(Celsius) for a primer oligo.<br>
 
<h4><a name="minprimerTm">Minimum primer Tm</a></h4>
Minimum acceptable melting temperature(Celsius) for a primer oligo.<br>
 
<h4><a name="maxprimerGC">Maximum primer GC%</a></h4>
Maximum allowable percentage of Gs and Cs in any primer.<br> 

<h4><a name="minprimerGC">Minimum primer GC%</a></h4>
Minimum allowable percentage of Gs and Cs in any primer.<br> 

<h4><a name="maxcom">Maximum complementarity</a></h4>
The maximum allowable local alignment score when testing a single primer for (local) self-complementarity
and the maximum allowable local alignment score when testing for complementarity between left and right
primers. Local self-complementarity is taken to predict the tendency of primers to anneal to each other
without necessarily causing self-priming in the PCR. Scores are non-negative, and a score of 0.00
indicates that there is no reasonable local alignment between two oligos.<br>

<h4><a name="max3com">Maximum 3' complementarity</a></h4>
The maximum allowable 3'-anchored global alignment score when testing a single primer for self-
complementarity, and the maximum allowable 3'-anchored global alignment score when testing for
complementarity between left and right primers. The 3'-anchored global alignment score is taken to 
predict the likelihood of PCR-priming primer-dimers. The scoring system is as for the Maximum
Complementarity argument. Scores are non-negative, and a score of 0.00 indicates that there is no
reasonable 3'-anchored global alignment between two oligos. It is nonsensical to provide a larger value
for this parameter than for the Maximum (local) Complementarity parameter because the score of a local
alignment will always be at least as great as the score of a global alignment.<br>

<h4><a name="saltconc">Salt concentration (mM)</a></h4>
The millimolar concentration of salt (usually KCl) in the PCR. The program uses this argument to
calculate oligo melting temperatures. Default is 50.0 mM.<br> 

<h4><a name="primerconc">Annealing primer concentration (mM)</a></h4>
The nanomolar concentration of annealing oligos in the PCR. This argument is used to calculate oligo
melting temperatures. The default value is set to 50 (nM). The value of the parameter is less than
the actual concentration of oligos in the reaction because it is the concentration of annealing oligos,
which in turn depends on the amount of template (including PCR product). This concentration increases
a great deal during a PCR. However, PCR seems quite robust for a variety of oligo melting temperatures.<br>

<h4><a name="numoutput">Number of outputs</a></h4>
The maximum number of primer pair combinations to return. Setting this parameter to a large value (upper
limit is 1000) will increase running time.<br> 

<br><br>
<a href="#top">Back to top</a><br>
</body>
</html>
 
